### PR TITLE
ARROW-11790: [Rust][DataFusion] RFC: Change builder signatures to take Vec<Expr> rather than &[Expr]

### PR DESCRIPTION
**Pull request by alamb**
# NOTE:
Since is a fairly major backwards incompatible change (many callsites need to be updated, though mostly mechanically), I have opened this PR with an example of how it would work for a few callsites, and wanted to gauge community interest before getting the whole thing working.

It is part of my overall plan to make the DataFusion optimizer more idiomatic and do much less copying [ARROW-11689](https://issues.apache.org/jira/browse/ARROW-11689)

# Rationale:
All callsites currently need an owned `Vec` (or equivalent) so they can pass in `&[Expr]` and then Datafusion copies all the `Expr`s. Many times the original `Vec<Expr>` is discarded immediately after use (I'll point out where this happens in a few places below). Thus I it would better (more idiomatic and often less copy/faster) to take something that could produce an iterator over Expr

# Changes
Change `Dataframe` / `LogicalPlanBuilder` so it takes `Vec<Expr>` rather than `&[Expr]`

### CODE REVIEW

File: rust/datafusion/src/dataframe.rs [hunk 1]
```diff
@@ -75,11 +75,11 @@ pub trait DataFrame: Send + Sync {
     /// # fn main() -> Result<()> {
     /// let mut ctx = ExecutionContext::new();
     /// let df = ctx.read_csv("tests/example.csv", CsvReadOptions::new())?;
-    /// let df = df.select(&[col("a") * col("b"), col("c")])?;
+    /// let df = df.select(vec![col("a") * col("b"), col("c")])?;
     /// # Ok(())
     /// # }
     /// ```
-    fn select(&self, expr: &[Expr]) -> Result<Arc<dyn DataFrame>>;
+    fn select(&self, expr: Vec<Expr>) -> Result<Arc<dyn DataFrame>>;
```
Comment (alamb) [inline rust/datafusion/src/dataframe.rs:82]: Here is the API change -- most of the rest of this PR is updating all the call sites. Note that [`DataFrame::filter`](https://github.com/apache/arrow/pull/9692/files#diff-b31224e33fb55c5ba87aecede66d7f74d7dd425020a47804e6a8028d4809a8e6R96) takes in an owned `Expr` so taking an owned `Vec` is not that large of a departure

File: rust/datafusion/src/logical_plan/builder.rs [hunk 2]
```diff
@@ -132,16 +132,19 @@ impl LogicalPlanBuilder {
...
-    pub fn project(&self, expr: &[Expr]) -> Result<Self> {
+    pub fn project(&self, expr: Vec<Expr>) -> Result<Self> {
         let input_schema = self.plan.schema();
         let mut projected_expr = vec![];
-        (0..expr.len()).for_each(|i| match &expr[i] {
-            Expr::Wildcard => {
-                (0..input_schema.fields().len())
-                    .for_each(|i| projected_expr.push(col(input_schema.field(i).name())));
-            }
-            _ => projected_expr.push(expr[i].clone()),
```
Comment (alamb) [inline rust/datafusion/src/logical_plan/builder.rs:143]: Here is where the `Expr::clone` occurs and no longer does after the changes contemplated in this PR

**Comment by alamb** [comment]
I would love your thoughts on this approach  @jorgecarleitao @andygrove @houqp @Dandandan @ilya-biryukov @seddonm1  -- I wonder if this approach would be worth bringing up on the mailing list

**Comment by houqp** [comment]
Since it aligns with our previous discussion at https://github.com/apache/arrow/pull/9347#pullrequestreview-581673922, I am all for this change! I think we should adopt this pattern across the entire code base. Clone should be managed at callsites whenever possible.

**Comment by andygrove** [comment]
I am all for this change. I had started out with the pattern of passing `Vec` but at the time I got feedback that this wasn't idiomatic Rust code and that we should always pass slices, so had switched to that pattern.

**Comment by houqp** [comment]
> this wasn't idiomatic Rust code and that we should always pass slices

I also remember seeing that recommendation before and adopted it for awhile. I wonder how wide-spread this is in the Rust community :thinking:

**Comment by Dandandan** [comment]
I think this makes sense. This is a partial revert from a PR from me (I believe). But I think it is better for "builder" kind of functions where the data should

**Comment by houqp** [comment]
> In general, &Vec and &String are anti patterns though, and some functions taking Vec and String still could/should be using references instead.

Yeah, I think it's always better to use `&[]` instead of `&Vec` and `&str` instead of `&String`.

In case of `Vec` v.s. `&[]`, I think it largely depends on the clone overhead. For example, if the function doesn't do any clone at all internally, then we should definitely pass in `&[]` instead of an owned `Vec`. If there is a decent amount of clone overhead, then it's better to manage the clone on callsite.

**Comment by Dandandan** [comment]
> > In general, &Vec and &String are anti patterns though, and some functions taking Vec and String still could/should be using references instead.
> 
> Yeah, I think it's always better to use `&[]` instead of `&Vec` and `&str` instead of `&String`.
> 
> In case of `Vec` v.s. `&[]`, I think it largely depends on the clone overhead. For example, if the function doesn't do any clone at all internally, then we should definitely pass in `&[]` instead of an owned `Vec`. If there is a decent amount of clone overhead, then it's better to manage the clone on callsite.

Exactly, this is my current understanding too. It is best to evaluate on a case by case basis 👍

**Review by jorgecarleitao — APPROVED** [review]
LGTM. Thanks a lot, @alamb !

I agree with what was said: when the function needs ownership, it is better to leave the caller to make the decision on whether they want to clone it and keep the object or pass the ownership to the function.

I do think that in some cases we could simplify some structs by having a lifetime parameter and `&'a []` instead of `Vec`, but since this is not high performant code, I do not think we need to worry much.

**Comment by alamb** [comment]
@Dandandan  and @houqp 

What do you think about something like this (which is even more Idiomatic (TM) I think), though harder for beginner Rust programmers to grok? 

```
    pub fn project(&self, expr: impl IntoIterator<Item=Expr>) -> Result<Self> {
```

This would allow `PlanBuilder::project` to be called with a `Vec<Expr>` as well as things like `BTreeSet<Expr>`...


Sadly, I couldn't give the same treatment to the `DataFrame` trait as when I tried,  `rustc` started complaining to me about trait objects

```
dataframe::DataFrame` cannot be made into an object
   |
   = help: consider moving `select` to another trait
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
```

**Comment by houqp** [comment]
Yes, I think if the function only requires sequential access, iterator would have been even better. Although I would prefer to write it as a generic function rather than using Trait object.

**Comment by jorgecarleitao** [comment]
Another option is to use `T: AsRef<[Expr]>`, which works for any type that can be represented as a reference to an array (e.g. `vec` and &[]`)

@houqp , AFAIK `impl` is just syntactic sugar for a generic; a trait object would be `&dyn Expr`.

I am not sure the `IntoIterator` would help, though: we would still need to `collect` it into a vector, which afaik results in a clone.

**Comment by houqp** [comment]
@jorgecarleitao yeah, you are right, sorry I got it mixed up with `dyn` :P

I think `IntoIterator<Item = Expr>` won't require copy or clone because the iterator owns those values. `AsRef` with slice on the other side would require copy/clone due to use of slice, which behaves closer to `IntoIterator<Item=&Expr>`.

**Comment by jorgecarleitao** [comment]
@houqp , you are right, `AsRef` does not work. 👍  [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5cbb014dde14a214d08bf5fce7675e69).

Thank you for that piece of knowledge! :)

**Comment by alamb** [comment]
Closing in favor of https://github.com/apache/arrow/pull/9703

**Pull request closed without merging**